packages/core/src/lib/create-prompt.mts

```diff
@@ -86,17 +81,23 @@ export function createPrompt<Value, Config>(view: ViewFunction<Value, Config>) {
       rl.on('close', hooksCleanup);
       cleanups.add(() => rl.removeListener('close', hooksCleanup));
 
-      function done(value: Value) {
-        // Delay execution to let time to the hookCleanup functions to registers.
-        setImmediate(() => {
-          onExit();
-
-          // Finally we resolve our promise
-          resolve(value);
-        });
-      }
-
       cycle(() => {
+        let isCycleDone = false;
+        let afterCycle: (() => void) | undefined;
+        function done(value: Value) {
+          if (isCycleDone) {
+            setImmediate(() => {
+              hooksCleanup();
+              resolve(value);
+            });
+          } else {
+            afterCycle = () => {
+              hooksCleanup();
+              resolve(value);
+            };
+          }
+        }
```

Comment:
Not super satisfied with this block of logic, might revisit.

Reply:
The tl;dr: if we don't cleanup the hooks as part of the cycle, we ignore/silence the errors thrown while cleaning up and end up with a resolved promise.

If we didn't have to surface hook errors, then the whole logic goes away and we simply end up with:

```diff
- const nextView = view(config, done);
+ const nextView = view(config, resolve);
```

But I think surfacing those errors is important 🤔

Reply:
Latest mixes some promises together to get to the right result and order of operation. I'll reflect more on it, but it might be an acceptable solution.